crates/transpiler/src/passes/commutative_optimization.rs

```diff
@@ -0,0 +1,516 @@
+                } else {
+                    return Ok((true, Some(merged_instruction), 0.));
+                }
+            } else {
+                return Ok((false, None, 0.));
+            }
+        }
+    }
+
+    // Matrix-based check: the product matrix is equivalent to identity.
+    if inst1.op.num_qubits() <= matrix_max_num_qubits {
+        let view1 = inst1.op.view();
+        let view2 = inst2.op.view();
+
+        if let (Some(matrix1), Some(matrix2)) = (
+            get_matrix(
+                &view1,
+                inst1.params_view(),
+                Some(matrix_max_num_qubits),
+                true,
+            ),
+            get_matrix(
+                &view2,
+                inst2.params_view(),
+                Some(matrix_max_num_qubits),
+                true,
+            ),
+        ) {
+            let product_mat = matrix1.dot(&matrix2);
+            let dim = product_mat.shape()[0] as f64;
+            let tr_over_dim = product_mat.diag().iter().sum::<Complex64>() / dim;
+
+            let (can_be_removed, phase_update) = can_remove(tr_over_dim, dim, tol);
+            if can_be_removed {
+                return Ok((true, None, phase_update));
+            }
+        }
+    }
+
+    // Could not merge.
+    Ok((false, None, 0.))
+}
+
+#[pyfunction]
+#[pyo3(name = "commutative_optimization")]
+#[pyo3(signature = (dag, commutation_checker, approximation_degree=1., matrix_max_num_qubits=0))]
+pub fn run_commutative_optimization(
+    dag: &mut DAGCircuit,
+    commutation_checker: &mut CommutationChecker,
+    approximation_degree: f64,
+    matrix_max_num_qubits: u32,
+) -> PyResult<Option<DAGCircuit>> {
+    let tol = 1e-12_f64.max(1. - approximation_degree);
+
+    let error_cutoff_fn = |_inst: &PackedInstruction| -> f64 { tol };
+
+    // Create output DAG.
+    // We will use it to intern qubits of canonicalized instructions.
+    // (In theory, we could also change qubits when merging instructions, however
+    // this does not happen right now).
+    let mut new_dag = dag.copy_empty_like(VarsMode::Alike)?;
+
+    let node_indices = dag.topological_op_nodes()?.collect::<Vec<_>>();
+    let num_nodes = node_indices.len();
+
+    let mut node_actions: Vec<NodeAction> = vec![NodeAction::Keep; num_nodes];
+    let mut new_global_phase = dag.global_phase().clone();
+
+    let mut modified: bool = false;
+
+    for idx1 in 0..num_nodes {
+        let node_index1 = node_indices[idx1];
+        let instr1 = dag[node_index1].unwrap_operation();
+
+        let (can_be_removed, phase_update) =
+            is_identity_equiv(instr1, Some(matrix_max_num_qubits), true, error_cutoff_fn)?;
+        if can_be_removed {
+            node_actions[idx1] = NodeAction::Drop;
+            new_global_phase = radd_param(new_global_phase, Param::Float(phase_update));
+            modified = true;
+            continue;
+        }
+
+        if let Some((new_instruction, phase_update)) = canonicalize(&mut new_dag, instr1) {
```

Comment:
Are the variables `new_instruction` and `phase_update` used? It doesn't seem like it.. 🤔

Reply:
We are using these variables on the line immediately below -- to create the new canonical instruction :).